Generating sitemaps for multiple sites

YKWeyer opened the pull request:
When you use the [multisite function of Pimcore](https://www.pimcore.org/wiki/pages/viewpage.action?pageId=14551657) the generated sitemap.xml has to be specific to each subsite.
Furthermore, it should include the domain of the site, and not the global domain (or Google will start to index pages on the wrong website).
## Summary

This PR provides the following changes:
- Generating one sitemap.xml for each available site
- Using a controller (and staticroute) to select the right sitemap to display depending on the current domain
- Notify Google for every updated sitemap

Its major limitation is I found no standardized way to know the protocol used for every subsite (http or https), it seems some developers use a custom property on the website but there is no way to infer it. I assumed all the websites should use https, since the admin interface is available on the same domain _by default_, and everywhere there is a login form we should use https. But if you know any way to know it for sure, i'd really like to remove [that static part of the code](https://github.com/byng-systems/pimcore-sitemap-plugin/pull/6/files#diff-71d0e7681cf36c448120f554e8c70f94R107).
## How to install:
- on a fresh install, this is a regular plugin installation
- if you already have the plugin installed, you have to remove the `sitemap.xml` file on your root installation to be sure the request is processed by pimcore, not served by the http server
## Feedback needed !

I am fairly new to Pimcore, so everything might not be 100% optimized. The documentation being sometimes sporadic, I suppose there are some hidden tricks I don't know about 😄 
Also, I tested it only with our Pimcore website, which runs under 3.1.1.

I used some PHP 5.4+ codestyle (ternary operators, short array notation, etc.) => I suppose this is no major issue, since [composer.json](https://github.com/byng-systems/pimcore-sitemap-plugin/blob/16bf606d48c4bfd1ebfed39f0f48b8e96eb503c2/composer.json) requires 5.5+ ?

Commit:
Add controller to send right sitemap depending on the current subset

Commit:
Use plugins subfolder instead of root

Commit:
Generate XML sitemap for each site

Commit:
Retrieve siteRequest from document (useful when proxying domains to subfolders)

Also add comments

Commit:
Rename parameter to be more explicit, and add some documentation

Commit:
Append protocol before hostUrl (needed to follow conventions conventions)

Commit:
Add todo concerning notifying Google for each subsite sitexml

Commit:
Notify searchEngines for all the subsites as well

seeruk commented:
Hi @YKWeyer, thanks for opening this PR. I'm not in the office this week, and my colleagues may be a little too busy to look just yet. If they've not had a chance to take a look before I'm back then I'll get straight onto it.

PHP 5.5 support is totally fine by the way! :)

david-byng commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -128,16 +168,18 @@ private function getDateFormat($date)
         $googleNotifier = new GoogleNotifier();
 
-        if ($googleNotifier->notify()) {
+        if ($googleNotifier->notify($domain = null)) {
```
1. assign `null` to `$domain`
2. pass `$domain` (now `null`) in as a parameter to `notify`

Did you mean to add the `= null` here?

YKWeyer replied:
Damn you're right, this is a mistake! I'll fix it, thanks for noticing it!

Commit:
fix: domain was always null when calling googleNotifier

david-byng commented:
This looks great! From what I can see there are just a few minor picky code-style issues left (missing spaces, braces on if-statements, and the `array()` syntax, that sort of thing). I'm more JS focused though, so I'll have to leave it to our Pimcore experts to push the button.

Thanks @YKWeyer!

seeruk commented on controllers/SitemapController.php:
```diff
@@ -0,0 +1,32 @@
+ * Pimcoresitemapplugin_SitemapController
+ *
+ * @author Yann Weyer
+ */
```
I think there'll probably be a few quick style things like this, but, we have a common file header across all of the PHP source files in the repository that's like this:

``` php
<?php

/**
 * This file is part of the pimcore-sitemap-plugin package.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
```

seeruk commented on controllers/SitemapController.php:
```diff
@@ -0,0 +1,32 @@
+
+use Byng\Pimcore\Sitemap\SitemapPlugin as SitemapPlugin;
+
+class Pimcoresitemapplugin_SitemapController extends \Pimcore\Controller\Action\Frontend
```
Can the `Frontend` class be `use`d at the top of the file, and can this class be made final please?

seeruk commented on controllers/SitemapController.php:
```diff
@@ -0,0 +1,32 @@
+class Pimcoresitemapplugin_SitemapController extends \Pimcore\Controller\Action\Frontend
+{
+
+    /*
```
I think PhpDoc requires `/**` to be the start of a doc block. Could `@return void` also be added here too please?

SeerUK commented on controllers/SitemapController.php:
```diff
@@ -0,0 +1,32 @@
+        readfile(PIMCORE_WEBSITE_PATH . SitemapPlugin::SITEMAP_FOLDER . $filename);
+        exit;
+    }
+}
```
We also require that all source files end with a blank new line.

cj123 commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -65,18 +66,51 @@ public function __construct()
+        foreach($this->sitesRoots as $siteRootID => $siteRootDomain){
+            $this->generateSiteXml($siteRootID, $siteRootDomain);
+
+            if ($notifySearchEngines)
```
could you wrap this if in braces?

cj123 commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -65,18 +66,51 @@ public function __construct()
+        );
+
+        // Set current hostUrl
+        $this->hostUrl = 'https://' . $hostUrl;
```
I know that in 2016 all sites should be using HTTPS, but is it right to assume it here?

YKWeyer replied:
Hey, as I explained in the summary, I found no way to get the protocol of the website in Pimcore settings 😕 

I thought of the following possible resolutions, but I'm not quite happy with any of them: 
- Creating a new variable in `Settings > Website` tab (assumes all the sites and subsites share the same protocol)
- create custom plugin config file to set parameters for each site? => XML would be generated at the installation, then the addition of new subsites has to be done manually
- try curl request to https version of the hosturl, use http if error is returned (then cache the outcome)

If you have any idea...

cj123 replied:
ah sorry, I jumped straight into the code! 😀  

Agreed, I think I've met this problem before. All of those solutions have fairly obvious limitations unfortunately.

I kinda like the cURL idea, but it does seem a bit hacky. @SeerUK, @asim-byng - any ideas?

seeruk replied:
I think a configuration file may be the best of the options there. Even with caching the curl result, Pimcore isn't always that speedy. Whilst website settings are indeed convenient this kind of configuration doesn't really need to be in the UI at all (keep it away from people who may potentially muck it up! :joy:).

YKWeyer replied:
Ultimately the best solution would be to handle those parameters natively in Pimcore. I don't think we are the only ones having this problem 😐  
I'll try the configuration option and let you know when I'll have something to show!

YKWeyer replied:
Alright, I tried the configuration file way here: https://github.com/byng-systems/pimcore-sitemap-plugin/pull/6/commits/98fb54b0bce524a8a9bfbfe467e6c6611e23a87b

At the installation, we list all the sites and try to reach the domains using https to determine the right protocol to use. This configuration is stored in an editable XML with other useful parameters (domain, id, path). This XML is then used to generate the sitemaps (instead of fetching the list of available sites every time).

The major drawback of this approach is every time new sites are added or site domains are changed, the XML has to be changed manually (or the admin should uninstall/reinstall the plugin). 
But it is definitely more flexible than presuming https will be used every time. Furthermore, it allows admins to choose `http` as favourite protocol even if their domain handles `https`.

Commit:
Make controller class final and use `use` operator

Commit:
Update documentation and formatting

Commit:
Wrap condition in curly braces

Commit:
Format code in SitemapGenerator.php

cj123 commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -65,18 +66,52 @@ public function __construct()
+    /**
+     * Generate a sitemap xml file for a defined site, with a specific hostUrl
+     *
+     * @param $rootId
```
could you add the type of these parameters into the docblock?

Commit:
Add missing braces and documentation

Commit:
Create and populate an XML file at installation

Commit:
Check the config file is available in the isInstalled function

Commit:
Create folder only if it doesn't exist already

It avoids throwing an error when reinstalling after improper uninstall

Commit:
Use config XML to generate subsites sitemaps

Commit:
Store sites config in indexed array to access it later

Commit:
Remove rootPath from fullPath when using subsites

When accessing the page /subsite/page on the domain www.subsite.com, the fullPath should be /page

Commit:
Store multisite settings (protocol, host, ...) in an editable config file  

The configuration is generated at installation, and can be manually edited.

Commit:
Quickfix: use SitemapFolder const in ConfigurationFile path

Commit:
Move the plugin installation logic in a dedicated class

Commit:
Use PHP constants rather than static class constants

YKWeyer commented:
Hey @SeerUK , I tidied up a bit the code:
- Move the (un-)install logic to a dedicated class
- Allow to override the default folder and configuration files in custom installations

Let me know whether you still need changes in the code, or more documentation?

Commit:
Use native SimpleXMLElement instead of Zend_Config

Commit:
Fix homepage not being included for subsites

seeruk commented:
Hey @YKWeyer, it's been a while! I've just taken a look over the code, and it's looking great. I'm just going to talk it over with some of the guys who are more actively using this in a couple of our projects to make sure it'll still work - if it's all fine, we should be able to get it merged.

Thank you for taking the time to extend this, but moreso for contributing back too!

ioannisGiak89 commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -36,26 +36,30 @@
     /**
      * @var SimpleXMLElement
```
This is not SimpleXMLElement

YKWeyer replied:
Could you be more specific?

```
// Retrieve site trees
$config = simplexml_load_file(SITEMAP_CONFIGURATION_FILE);
$siteRoots = $config->sites->site;
```
=> The site parameter is an XML block containing several tags. I think `SimpleXMLElement` is the right type here, but I'm open to other alternatives

YKWeyer replied:
Hi @ioannisGiak89 let me know how I can help :)

ioannisGiak89 replied:
I mean the dockblock here is for the `private $host` field and not for the `private $xml` field which is further down. Sorry if it wasn't clear
`

    /**
     * @var SimpleXMLElement
     */
    private $host;
    /**
     * @var SimpleXMLElement
     */
    private $xml;

`

YKWeyer replied:
Thank you for your feedback! Actually it still should be a `SimpleXMLElement`, but I completely understand why this is a bit puzzling. This is something I improved in my last commits, without knowing you were reviewing an older version. I do think it should answer your interrogation (you might want to check the [GenerateSitemapXML method](https://github.com/byng-systems/pimcore-sitemap-plugin/pull/6/commits/96c8cfe49a7a271e59dc5cfbcaf821a3e00fa275#diff-71d0e7681cf36c448120f554e8c70f94R97) for example). 

Let me know if you think it still needs work!

seeruk replied:
This looks fine to me :)

ioannisGiak89 replied:
You are right!! Sorry for the confusion

ioannisGiak89 commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -36,26 +36,30 @@
+    /**
+     * @var array
+     */
+    private $sitesRoots = array();
```
Wouldn't be better as []?

YKWeyer replied:
You're right 👍

YKWeyer replied:
Fixed in commit b4dac46, let me know if you need anything else

ioannisGiak89 commented on lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php:
```diff
@@ -107,9 +149,22 @@ private function addUrlChild(Document $document)
+                }
+            }
+
+            echo $this->hostUrl . $fullPath . "\n";
```
Do you need to echo anything?

YKWeyer replied:
I don't think we need it, but it was [in the original method](https://github.com/byng-systems/pimcore-sitemap-plugin/blob/master/lib/Byng/Pimcore/Sitemap/Generator/SitemapGenerator.php#L110), I just kept it.

Commit:
Restric redirect rule to exact string

Commit:
Improve code structure and comments

Commit:
Use short notation for array creation

seeruk commented:
Alrighty, I've _FINALLY_ been able to test this thing out properly. I've only noticed one thing, which is that the default sitemap _may_ break. If you've not got a domain set, it just makes a file for it called `.xml`, maybe it should fall back to `sitemap.xml` and use that only for the main site?

Now I've got this thing to test it set up, it should be faster to get this thing closed once that change is made!

YKWeyer commented:
Hi @SeerUK , thank you for your feedback!
I am afraid we need a domain to be set anyway since a sitemap.xml requires absolute URLs. Even if we fall back to `sitemap.xml` if no domain is set, what would you use as a hostUrl (and protocol)?
Worst-case solution could be to generate the sitemap on the fly for the current host if no file match the current request?

seeruk commented:
Hey @YKWeyer, generating it on the fly might be too taxing... I don't think it should generate the file if there isn't a domain set, and it should just 404 or something in the controller if there isn't a sitemap.xml file too?

Commit:
Throw an exception if the sitemap file isn't found

YKWeyer commented:
@SeerUK The controller now checks if the appropriate file exists, and throws a 404 error if not.

Commit:
Fix config path still using capitalised folder

seeruk commented on lib/Byng/Pimcore/Sitemap/Plugin/Installer.php:
```diff
@@ -0,0 +1,201 @@
+    public function deleteSitemapFolder()
+    {
+        if (is_dir(SITEMAP_PLUGIN_FOLDER)) {
+            if (!is_dir_empty(SITEMAP_PLUGIN_FOLDER)) {
```
I'm getting an error here, saying `is_dir_empty` doesn't exist, causing the uninstall function to fail.

YKWeyer replied:
My bad, this is an utility function that has been added only for Pimcore 4.x versions and up. I'll replace it!

seeruk commented on lib/Byng/Pimcore/Sitemap/Plugin/Installer.php:
```diff
@@ -0,0 +1,201 @@
+            'rootPath' => '',
+            'protocol' => $this->getProtocolForDomain($defaultDomain, $client),
+            'domain' => $defaultDomain
+        ];
```
Can the above array be wrapped in an if that checks if `$defaultDomain` is truthy? Otherwise it also errors here when installing if you've not set a general domain. I'd either do that, or make it so that it shows you a more specific error message telling you that you need to add that domain.

seeruk commented:
I feel like we owe you a drink after this. It's been a long road, but I can see the end in sight!

seeruk reviewed (changes requested):
Couple of comments related to installation / uninstallation - but I think that'll be it!

Commit:
Improve installer logic based on last review

YKWeyer commented:
@SeerUK I finally got some time to implement those minor changes.

> Can the above array be wrapped in an if that checks if $defaultDomain is truthy? Otherwise it also errors here when installing if you've not set a general domain. I'd either do that, or make it so that it shows you a more specific error message telling you that you need to add that domain.

I implemented the simplest solution. If there is no domain set, then the main domains sitemap won't be generated. I don't know how many instances it might affect.
A simple, seamless fallback would be to use the admin domain by default, but once again, I have no idea of the most frequent configurations/ use cases.

YKWeyer commented:
I'm planning on archiving/removing my fork at the end of this month. 

Since this repository hasn't been updated since 2016, I assume it is not maintained anymore. But I wanted to give you notice before I remove it.

Even if it wasn't merged at the end, I really enjoyed the discussion. Thank you all for taking the time to review this PR :pray: :hugs:
